src/darwin/Framework/CHIP/MTRDevice_Internal.h

```diff
@@ -65,8 +66,57 @@ MTR_TESTABLE
 - (nullable instancetype)initWithDataVersion:(NSNumber * _Nullable)dataVersion attributes:(NSDictionary<NSNumber *, MTRDeviceDataValueDictionary> * _Nullable)attributes;
 @end
 
-@interface MTRDevice ()
-- (instancetype)initForSubclasses;
+// Consider moving utility classes to their own file
+#pragma mark - Utility Classes
+
+/**
+ * container of MTRDevice delegate weak reference, its queue, and its interested
+ * paths for attribute reports.
+ */
+MTR_DIRECT_MEMBERS
+@interface MTRDeviceDelegateInfo : NSObject {
+@private
+    void * _delegatePointerValue;
+    __weak id _delegate;
+    dispatch_queue_t _queue;
+    NSArray * _Nullable _interestedPathsForAttributes;
+    NSArray * _Nullable _interestedPathsForEvents;
+}
+
+// Array of interested cluster paths, attribute paths, or endpointID, for attribute report filtering.
+@property (readonly, nullable) NSArray * interestedPathsForAttributes;
+
+// Array of interested cluster paths, attribute paths, or endpointID, for event report filtering.
+@property (readonly, nullable) NSArray * interestedPathsForEvents;
+
+// Expose delegate
+@property (readonly) id delegate;
+
+// Pointer value for logging purpose only
+@property (readonly) void * delegatePointerValue;
+
+- (instancetype)initWithDelegate:(id<MTRDeviceDelegate>)delegate queue:(dispatch_queue_t)queue interestedPathsForAttributes:(NSArray * _Nullable)interestedPathsForAttributes interestedPathsForEvents:(NSArray * _Nullable)interestedPathsForEvents;
+
+// Returns YES if delegate and queue are both non-null, and the block is scheduled to run.
+- (BOOL)callDelegateWithBlock:(void (^)(id<MTRDeviceDelegate>))block;
+
+#ifdef DEBUG
+// Only used for unit test purposes - normal delegate should not expect or handle being called back synchronously.
+- (BOOL)callDelegateSynchronouslyWithBlock:(void (^)(id<MTRDeviceDelegate>))block;
+#endif
+@end
+
+#pragma mark - MTRDevice internal extensions
+
+@interface MTRDevice () {
+    // Ivars needed to implement shared MTRDevice functionality.
+@protected
+    // Lock that protects overall device state, including delegate storage.
+    os_unfair_lock _lock;
+    NSMutableSet<MTRDeviceDelegateInfo *> * _delegates;
```

Comment:
Why?  I would genuinely like to understand what the benefits of that would be.  The main drawback is that making these properties makes it harder to reason about the code, since anyone outside the class can mess with the values (e.g. mutate the set while not locked, or take/release the lock).

In general, these are private state, not public state; why would they be properties?

Reply:
Looks like we have to use `@package` here for now, which is similar-ish to just having properties...  I'll look for a better way to do this in a followup.